src/Config/ConfigurationLoader.php

```diff
@@ -149,12 +148,59 @@ private function fetchAndStoreBandits(): void
         $banditModelVersions = array_map(fn($bandit)=> $bandit->modelVersion, $bandits);
 
         $this->configurationStore->setBandits($bandits);
-        $this->configurationStore->setMetadata(self::BANDIT_MODEL_VERSIONS_KEY, $banditModelVersions);
-        $this->configurationStore->setMetadata(self::BANDIT_TIMESTAMP, time());
+        $this->configurationStore->setMetadata(self::KEY_LOADED_BANDIT_VERSIONS, $banditModelVersions);
+        $this->configurationStore->setMetadata(self::KEY_BANDIT_TIMESTAMP, time());
     }
 
     public function getBandit(string $banditKey): ?Bandit
     {
         return $this->configurationStore->getBandit($banditKey);
     }
+
+    /**
+     * Loads bandits unless `optimizedBanditLoading` is `true` in which case, currently loaded bandit models are
+     * compared to those required by flags to determine whether to (re)load bandit models.
+     *
+     * @param IBanditReferenceIndexer $indexer
+     * @return void
+     * @throws HttpRequestException
+     * @throws InvalidApiKeyException
+     * @throws InvalidConfigurationException
+     */
+    private function fetchBanditsAsRequired(IBanditReferenceIndexer $indexer): void
+    {
+        if ($this->optimizedBanditLoading) {
+            // Get the currently loaded bandits to determine if they satisfy what's required by the flags
+            $currentlyLoadedBanditModels = $this->configurationStore->getMetadata(
+                self::KEY_LOADED_BANDIT_VERSIONS
+            ) ?? [];
+            $references = $indexer->getBanditModelVersionReferences();
+
+            $needToFetch = !self::satisfiesRequiredModels($currentlyLoadedBanditModels, $references);
+        } else {
+            // Not optimized loading.
+            $needToFetch = true;
+        }
+
+        if ($needToFetch) {
+            $this->fetchAndStoreBandits();
+
+            // What happens if the newly loaded models don't match ??!!?
+        }
+    }
+
+    /**
+     * @param array<string, string> $loaded
+     * @param array<string, string> $required
+     * @return bool
+     */
+    private static function satisfiesRequiredModels(array $loaded, array $required): bool
```

Comment:
This logic is fine, but. I don't think we need to compare maps and instead could just compare sets. Basically, does the set of loaded versions contain all of the required versions.

Then you could get rid of the sub-function and simply do:
```php
$needToFetch = !!array_diff($loaded, $required);
```